File: Buffer/frame_tags.go
// Buffer/frame_tags.go — the fd-3 frame ENVELOPE discriminator.
//
// This is deliberately NOT part of the generated column-layout schema
// (Buffer/layout.go / buffer_layout_gen.go / buffer-layout.ts): it is not a column
// inside a block, it is the outer frame's own tag byte, sitting alongside the u32
// length prefix. It is hand-authored here and mirrored by hand in
// tools/topology-vscode/src/schema/frame-tags.ts — mirroring the existing
// input_codec.go / input-layout.ts split (binary wire ENVELOPE constants live
// hand-paired, not generated).
//
// Frame format on fd 3: [len:u32-LE][blockTag:u8][block bytes] where len counts the
// tag byte plus the block bytes. Four tag values exist today:
//
//   - BufBlockTagScene: the combined snapshot (layout-link/camera/overlay/scene/event
//     blocks — see snapshot.go), built by buildSnapshot. No longer carries beads
//     (BufBlockTagBead), the node-owner-group blocks (BufBlockTagNode), or the Edge
//     block + edge-label bytes (BufBlockTagEdge).
//
//   - BufBlockTagBead: the Bead block ALONE, in its own self-contained frame, built by
//     buildBeadFrame (Buffer/pack.go). Beads churn every tick, independent of the rest
//     of scene state, so they are split into their own per-tick frame rather than
//     riding the scene frame. Its payload layout is:
//
//     BufBeadHeaderSize (8) bytes: [tick:u32][beadCount:u32]
//     Bead                  beadCount × BufBeadStride bytes (same Bead block columns
//     as before — Buffer/layout.go's bufLayoutBead — unchanged)
//
//   - BufBlockTagNode: the Node/Interior/Port blocks + their string sections (node Label
//     bytes, Port-name bytes) ALONE, in their own self-contained frame, built by
//     buildNodeFrame (Buffer/pack.go). These three blocks share ONE owner group (the
//     node movers), so they travel together, split out of the scene frame the same way
//     the Bead block was. Its payload layout is:
//
//     BufNodeFrameHeaderSize (20) bytes: [tick:u32][nodeCount:u32][portCount:u32]
//     [labelBytesCount:u32][portNameBytesCount:u32]
//     Node      nodeCount × BufNodeStride bytes
//     Interior  nodeCount × BufInteriorSlotsPerNode × BufInteriorStride bytes (fixed
//     slots per node — no separate count; length derives from nodeCount)
//     Port      portCount × BufPortStride bytes (flattened over nodes in node-row order)
//     Label     labelBytesCount bytes (node labels' UTF-8 bytes, node-row order)
//     PortName  portNameBytesCount bytes (port names' UTF-8 bytes, flattened port-row order)
//
//     The scene frame's LayoutLink block still packs SrcNodeRow/DstNodeRow as node-row
//     indices (nodeRowIndex) — those rows resolve against THIS frame's Node block, since
//     both frames are built from the same SnapshotState in the same emitSnapshot call and
//     share the same stable node-row order (node id insertion order).
//
//   - BufBlockTagEdge: the Edge block ALONE + its EdgeLabel bytes section, in its own
//     self-contained frame, built by buildEdgeFrame. The Edge block carries NO endpoint
//     coordinates — see bufLayoutEdge's doc comment in layout.go for why (the endpoint-
//     tear fix): it references its two port rows (SrcPortRow/DstPortRow), which resolve
//     against the NODE frame's Port block (same reasoning as LayoutLink's SrcNodeRow/
//     DstNodeRow above — both frames share one SnapshotState/emitSnapshot call). Its
//     payload layout is:
//
//     BufEdgeFrameHeaderSize (12) bytes: [tick:u32][edgeCount:u32][edgeLabelBytesCount:u32]
//     Edge      edgeCount × BufEdgeStride bytes
//     EdgeLabel edgeLabelBytesCount bytes (edge labels' UTF-8 bytes, edge-row order)
//
//   - BufBlockTagView: the VIEW stream's own frame (camera + overlay + scene-sphere),
//     built by buildViewFrame. This is the first stream migrated OFF fd 3 onto its own
//     dedicated inherited pipe (see StreamFDs / main.go), per the no-single-writer-bridge
//     rule (memory/feedback_no_single_writer_bridge.md): the view/gesture goroutine gets
//     its own binary channel to TS instead of riding the fd-3 scene frame. When the
//     dedicated view fd is ACTIVE, the wire bytes on THAT fd carry NO tag byte — the fd
//     POSITION already identifies the stream (see StreamFDs' doc comment) — so
//     BufBlockTagView is never written to the wire in that mode; it exists only as the
//     synthetic tag the ext host attaches when relaying a decoded view frame onward to
//     the webview (mirrored in frame-tags.ts), so the webview's existing
//     tag-routed-cell pattern (scene/bead/node/edge) extends uniformly to a fifth cell
//     without a second message shape. When the dedicated view fd is NOT active (env
//     unset — headless tests, non-extension launch), camera/overlay/scene stay embedded
//     in the fd-3 SCENE frame exactly as before (the fallback path) and BufBlockTagView
//     is unused. Its payload layout (dedicated-fd wire bytes, no tag):
//
//     BufViewFrameHeaderSize (4) bytes: [tick:u32]
//     Camera   BufCameraStride bytes
//     Overlay  BufOverlayStride bytes
//     Scene    BufSceneStride bytes
//
// This is the protocol foundation for eventually splitting the rest of the single
// content buffer into N per-block buffers, each streamed as its own tagged frame; Bead,
// Node, and Edge above are three of that series; View is the first to move onto its own
// dedicated fd rather than just its own tag within fd 3 (do not add a further tag value
// to the fd-3 vocabulary until the next such split actually lands).
package Buffer

// BufBlockTagScene is the fd-3 block tag for the combined content-buffer snapshot
// (everything except beads, the node-owner-group blocks, and the Edge block), built by
// buildSnapshot.
const BufBlockTagScene byte = 0

// BufBlockTagBead is the fd-3 block tag for the self-contained per-tick Bead frame,
// built by buildBeadFrame. See this file's header comment for its payload layout.
const BufBlockTagBead byte = 1

// BufBlockTagNode is the fd-3 block tag for the self-contained Node/Interior/Port frame
// (+ Label/PortName bytes), built by buildNodeFrame. See this file's header comment for
// its payload layout.
const BufBlockTagNode byte = 2

// BufBlockTagEdge is the fd-3 block tag for the self-contained Edge frame (+ EdgeLabel
// bytes), built by buildEdgeFrame. See this file's header comment for its payload layout.
const BufBlockTagEdge byte = 3

// BufBeadHeaderSize is the byte width of the Bead frame's own header: [tick:u32][beadCount:u32].
// Hand-authored here (envelope-level, like the rest of this file) rather than generated,
// mirroring BufHeaderSize's split from the generated column-layout schema.
const BufBeadHeaderSize = 8

// BufNodeFrameHeaderSize is the byte width of the Node frame's own header:
// [tick:u32][nodeCount:u32][portCount:u32][labelBytesCount:u32][portNameBytesCount:u32].
// Hand-authored here (envelope-level, like the rest of this file) rather than generated,
// mirroring BufBeadHeaderSize/BufHeaderSize's split from the generated column-layout schema.
const BufNodeFrameHeaderSize = 20

// BufEdgeFrameHeaderSize is the byte width of the Edge frame's own header:
// [tick:u32][edgeCount:u32][edgeLabelBytesCount:u32]. Hand-authored here (envelope-level,
// like the rest of this file) rather than generated, mirroring
// BufNodeFrameHeaderSize/BufBeadHeaderSize/BufHeaderSize's split from the generated
// column-layout schema.
const BufEdgeFrameHeaderSize = 12

// BufBlockTagView is the SYNTHETIC ext-host-side tag for a decoded VIEW-stream frame,
// relayed to the webview under the same "buffer-snapshot" message shape as the fd-3
// tags above so the webview's existing tag-routed-cell pattern extends to a fifth cell.
// It is NEVER written as a wire tag byte on the dedicated view fd (that fd's frames
// carry no tag byte at all — see this file's header comment) and is unrelated to the
// fd-3 tag vocabulary (BufBlockTagScene/Bead/Node/Edge, values 0-3); it exists purely so
// Go and TS agree on ONE numeric value for "this is the view frame" when the ext host
// relays it. Mirrored by hand in frame-tags.ts's BUF_BLOCK_TAG_VIEW.
const BufBlockTagView byte = 4

// BufViewFrameHeaderSize is the byte width of the VIEW stream's own frame header on its
// dedicated fd: [tick:u32]. Hand-authored here (envelope-level), mirroring
// BufEdgeFrameHeaderSize/BufNodeFrameHeaderSize/BufBeadHeaderSize/BufHeaderSize's split
// from the generated column-layout schema.
const BufViewFrameHeaderSize = 4

// BufBlockTagEdgeStream is the SYNTHETIC ext-host-side tag for a decoded per-edge stream
// frame (see Buffer/stream_fds.go's StreamKindEdge / Buffer/edge_stream_frame.go), relayed
// to the webview under the same "buffer-snapshot" message shape as BufBlockTagView, PLUS a
// `row` field (the edge's stable row) so the webview can route it to the right per-edge
// cell — there are many edge streams (one per edge), unlike VIEW's singleton row. NEVER a
// wire tag byte: the dedicated per-edge fd's frames carry no tag byte at all (the fd
// POSITION already identifies which edge — see stream_fds.go). Mirrored by hand in
// frame-tags.ts's BUF_BLOCK_TAG_EDGE_STREAM.
const BufBlockTagEdgeStream byte = 5

// BufBlockTagNodeStream is the SYNTHETIC ext-host-side tag for a decoded per-node stream
// frame (see Buffer/stream_fds.go's StreamKindNode / Buffer/node_stream_frame.go's
// BuildNodeStreamFrame), relayed under the same "buffer-snapshot" shape as
// BufBlockTagEdgeStream, plus a `row` field (the node's stable row). NEVER a wire tag byte:
// the dedicated per-node fd's frames carry no tag byte at all (the fd POSITION already
// identifies which node). Mirrored by hand in frame-tags.ts's BUF_BLOCK_TAG_NODE_STREAM.
const BufBlockTagNodeStream byte = 6

// BufBlockTagInteriorStream is the SYNTHETIC ext-host-side tag for a decoded per-node
// INTERIOR stream frame (see Buffer/stream_fds.go's StreamKindInterior /
// Buffer/node_stream_frame.go's BuildInteriorStreamFrame), relayed under the same shape as
// BufBlockTagNodeStream, plus a `row` field (same node-row numbering). NEVER a wire tag
// byte. Mirrored by hand in frame-tags.ts's BUF_BLOCK_TAG_INTERIOR_STREAM.
const BufBlockTagInteriorStream byte = 7

File: tools/topology-vscode/src/schema/frame-tags.ts
// frame-tags.ts — hand-authored mirror of Buffer/frame_tags.go's fd-3 frame ENVELOPE
// discriminator. NOT generated (see that file's header comment for why: it's the
// outer frame's tag byte, not a column-layout block). Keep these values in lockstep
// with BufBlockTagScene/BufBlockTagBead by hand, the same way input-layout.ts mirrors
// input_codec.go.
//
// Frame format on fd 3 (see runCommand.ts splitFrames/handleFd3):
//   [len:u32-LE][blockTag:u8][block bytes]
// len counts the tag byte plus the block bytes. Four tag values exist today:
//
//   - BUF_BLOCK_TAG_SCENE: the combined snapshot (layout-link/camera/overlay/scene/event
//     blocks). No longer carries beads, the node-owner-group blocks, or the Edge block +
//     edge-label bytes (see BUF_HEADER_SIZE's comment in buffer-layout.ts).
//   - BUF_BLOCK_TAG_BEAD: the Bead block ALONE, in its own self-contained per-tick
//     frame. Its payload layout is:
//
//     BUF_BEAD_HEADER_SIZE (8) bytes: [tick:u32][beadCount:u32]
//     Bead                    beadCount × BEAD_STRIDE bytes (same Bead block columns
//                             as before — buffer-layout.ts's BEAD_* — unchanged)
//
//   - BUF_BLOCK_TAG_NODE: the Node/Interior/Port blocks + their string sections (node
//     Label bytes, Port-name bytes) ALONE, in their own self-contained frame. These
//     three blocks share ONE owner group (the node movers), so they travel together.
//     Its payload layout is:
//
//     BUF_NODE_FRAME_HEADER_SIZE (20) bytes: [tick:u32][nodeCount:u32][portCount:u32]
//       [labelBytesCount:u32][portNameBytesCount:u32]
//     Node      nodeCount × NODE_STRIDE bytes
//     Interior  nodeCount × INTERIOR_SLOTS_PER_NODE × INTERIOR_STRIDE bytes (fixed slots
//               per node — no separate count; length derives from nodeCount)
//     Port      portCount × PORT_STRIDE bytes (flattened over nodes in node-row order)
//     Label     labelBytesCount bytes (node labels' UTF-8 bytes, node-row order)
//     PortName  portNameBytesCount bytes (port names' UTF-8 bytes, flattened port-row order)
//
//     The scene frame's LayoutLink block still packs SrcNodeRow/DstNodeRow as node-row
//     indices — those rows resolve against THIS frame's Node block (both frames are
//     built from the same Go SnapshotState in the same emitSnapshot call and share the
//     same stable node-row order).
//
//   - BUF_BLOCK_TAG_EDGE: the Edge block ALONE + its EdgeLabel bytes section, in its own
//     self-contained frame. The Edge block carries NO endpoint coordinates: it
//     references its two port rows (SrcPortRow/DstPortRow), which resolve against THIS
//     SAME TICK's NODE frame's Port block — the endpoint's world position lives ONLY
//     there (node-owned), so a fast drag can never composite a fresh Node frame against
//     a stale Edge-block endpoint (the tear this replaces). Its payload layout is:
//
//     BUF_EDGE_FRAME_HEADER_SIZE (12) bytes: [tick:u32][edgeCount:u32][edgeLabelBytesCount:u32]
//     Edge      edgeCount × EDGE_STRIDE bytes
//     EdgeLabel edgeLabelBytesCount bytes (edge labels' UTF-8 bytes, edge-row order)
//
//   - BUF_BLOCK_TAG_VIEW: SYNTHETIC ext-host-side tag for a decoded VIEW-stream frame
//     (camera + overlay + scene-sphere), the first stream migrated OFF fd 3 onto its
//     own dedicated inherited pipe (see runCommand.ts's stream-fd allocation and
//     Buffer/stream_fds.go — memory/feedback_no_single_writer_bridge.md). The wire bytes
//     on that dedicated fd carry NO tag byte (the fd POSITION identifies the stream);
//     this tag exists only so the ext host can relay a decoded view frame to the
//     webview under the SAME "buffer-snapshot" message shape as the fd-3 tags above,
//     extending the existing tag-routed-cell pattern to a fifth cell instead of adding a
//     second message shape. Its payload layout (dedicated-fd wire bytes, no tag):
//
//     BUF_VIEW_FRAME_HEADER_SIZE (4) bytes: [tick:u32]
//     Camera   CAMERA_STRIDE bytes
//     Overlay  OVERLAY_STRIDE bytes
//     Scene    SCENE_STRIDE bytes
//
// This is the protocol foundation for eventually splitting the rest of the single
// content buffer into N per-block buffers, each streamed as its own tagged frame; Bead,
// Node, and Edge above are three of that series; View is the first to move onto its own
// dedicated fd rather than just its own tag within fd 3 (do not add a further tag value
// to the fd-3 vocabulary until the next such split actually lands).

/** The fd-3 block tag for the combined content-buffer snapshot (everything except beads,
 * the node-owner-group blocks, and the Edge block). */
export const BUF_BLOCK_TAG_SCENE = 0;

/** The fd-3 block tag for the self-contained per-tick Bead frame. See this file's header
 * comment for its payload layout. */
export const BUF_BLOCK_TAG_BEAD = 1;

/** The fd-3 block tag for the self-contained Node/Interior/Port frame (+ Label/PortName
 * bytes). See this file's header comment for its payload layout. */
export const BUF_BLOCK_TAG_NODE = 2;

/** The fd-3 block tag for the self-contained Edge frame (+ EdgeLabel bytes). See this
 * file's header comment for its payload layout. */
export const BUF_BLOCK_TAG_EDGE = 3;

/** Byte width of the Bead frame's own header: [tick:u32][beadCount:u32]. Hand-authored
 * (envelope-level), mirroring BUF_HEADER_SIZE's split from the generated column-layout schema. */
export const BUF_BEAD_HEADER_SIZE = 8;

/** Byte width of the Node frame's own header:
 * [tick:u32][nodeCount:u32][portCount:u32][labelBytesCount:u32][portNameBytesCount:u32].
 * Hand-authored (envelope-level), mirroring BUF_BEAD_HEADER_SIZE/BUF_HEADER_SIZE's split
 * from the generated column-layout schema. */
export const BUF_NODE_FRAME_HEADER_SIZE = 20;

/** Byte width of the Edge frame's own header: [tick:u32][edgeCount:u32][edgeLabelBytesCount:u32].
 * Hand-authored (envelope-level), mirroring BUF_NODE_FRAME_HEADER_SIZE/BUF_BEAD_HEADER_SIZE/
 * BUF_HEADER_SIZE's split from the generated column-layout schema. */
export const BUF_EDGE_FRAME_HEADER_SIZE = 12;

/** SYNTHETIC ext-host-side tag for a decoded VIEW-stream frame (camera+overlay+scene),
 * relayed to the webview under the same message shape as the fd-3 tags. NEVER a wire
 * tag byte on the dedicated view fd itself — see this file's header comment. Mirrors
 * Buffer/frame_tags.go's BufBlockTagView. */
export const BUF_BLOCK_TAG_VIEW = 4;

/** Byte width of the VIEW stream's own frame header on its dedicated fd: [tick:u32].
 * Hand-authored (envelope-level), mirroring BUF_EDGE_FRAME_HEADER_SIZE/.../BUF_HEADER_SIZE's
 * split from the generated column-layout schema. */
export const BUF_VIEW_FRAME_HEADER_SIZE = 4;

/** SYNTHETIC ext-host-side tag for a decoded per-edge dedicated-stream frame (one edgeMover
 * writes ITS OWN combined edge+bead frame to its own fd — see runCommand.ts's edge-fd
 * range and Buffer/stream_fds.go's StreamKindEdge). NEVER a wire tag byte on the dedicated
 * per-edge fd itself (the fd POSITION already identifies which edge). Relayed to the
 * webview under the same "buffer-snapshot" shape as BUF_BLOCK_TAG_VIEW, plus a `row` field
 * (there are many edge streams, not a singleton). Mirrors Buffer/frame_tags.go's
 * BufBlockTagEdgeStream. */
export const BUF_BLOCK_TAG_EDGE_STREAM = 5;

/** Byte layout of one edge's combined per-fd frame (Buffer.BuildEdgeStreamFrame), no outer
 * tag: [tick:u32] + one EDGE_STRIDE row (SrcPortRow/DstPortRow/Selected, EdgeLabelOff=0/Len)
 * + that edge's own label bytes (labelLen, from the row) + [beadCount:u32] + beadCount ×
 * BEAD_STRIDE bead rows. Header before the Edge row is just the tick. */
export const BUF_EDGE_STREAM_FRAME_HEADER_SIZE = 4;

/** SYNTHETIC ext-host-side tag for a decoded per-node dedicated-stream frame (one
 * nodeMover writes ITS OWN node geometry + ports + label to its own fd — see
 * runCommand.ts's node-fd range and Buffer/stream_fds.go's StreamKindNode). NEVER a wire
 * tag byte on the dedicated per-node fd itself (the fd POSITION already identifies which
 * node). Relayed to the webview under the same "buffer-snapshot" shape as
 * BUF_BLOCK_TAG_EDGE_STREAM, plus a `row` field (one per node row). Mirrors
 * Buffer/frame_tags.go's BufBlockTagNodeStream (ext-host-only; not carried on any Go wire
 * byte — this numbering only needs to stay distinct from the other synthetic tags here). */
export const BUF_BLOCK_TAG_NODE_STREAM = 6;

/** SYNTHETIC ext-host-side tag for a decoded per-node dedicated INTERIOR-stream frame (that
 * node's OWN Update goroutine writes its interior beads to its own fd — see
 * runCommand.ts's interior-fd range and Buffer/stream_fds.go's StreamKindInterior). NEVER a
 * wire tag byte on the dedicated fd itself. Relayed under the same "buffer-snapshot" shape,
 * plus a `row` field (one per node row, same numbering as BUF_BLOCK_TAG_NODE_STREAM). */
export const BUF_BLOCK_TAG_INTERIOR_STREAM = 7;

/** Byte layout of one node's combined per-fd frame (Buffer.BuildNodeStreamFrame), no outer
 * tag: [tick:u32][portCount:u32][labelLen:u32][portNameBytesCount:u32] + one NODE_STRIDE row
 * (LabelOff=0 into this frame's own label bytes) + labelLen label bytes + portCount ×
 * PORT_STRIDE port rows (each row's NodeRow column already the global node row) +
 * portNameBytesCount port-name bytes. */
export const BUF_NODE_STREAM_FRAME_HEADER_SIZE = 16;

/** Byte layout of one node's INTERIOR per-fd frame (Buffer.BuildInteriorStreamFrame), no
 * outer tag: [tick:u32] followed by a FIXED INTERIOR_SLOTS_PER_NODE × INTERIOR_STRIDE bytes
 * (no count — the decoder derives the length from the fixed per-node slot count, same as
 * the shared fd-3 Interior block). */
export const BUF_INTERIOR_STREAM_FRAME_HEADER_SIZE = 4;
